File: vscode-extension/src/verificationBehaviors.ts
import { TimestampedMessage } from './parser'
import { ParsedConversation } from './conversation'

const WINDOW = 10
const MIN_CONVS_FOR_RATE = 5
const EDIT_TOOLS = new Set(['Edit', 'Write', 'NotebookEdit'])
const REVIEW_TOOLS = new Set(['Read', 'Grep'])
const TEST_PATTERN = /\b(npm (run )?test|pytest|jest|cargo test|go test|rspec|mocha|vitest|bun test|tox)\b/i
const COMMIT_PATTERN = /\bgit commit\b/

export interface ConversationVerification {
  edits_count: number             // count of edit clusters (maximal Edit/Write runs)
  reviewed_edits: number          // clusters followed by Read/Grep within WINDOW tool positions
  commits_count: number           // count of `git commit` Bash invocations
  tested_commits: number          // commits with a test command in the prior WINDOW tool positions
  review_before_accept_rate: number  // reviewed_edits / edits_count, 0-1
  test_before_commit_rate: number    // tested_commits / commits_count, 0-1
}

export interface AggregateVerification {
  total_edit_clusters: number
  total_reviewed_edits: number
  review_before_accept_rate: number
  total_commits: number
  total_tested_commits: number
  test_before_commit_rate: number
  conversations_with_edits: number
  conversations_with_commits: number
  conversation_count: number
  insufficient_edits: boolean       // conversations_with_edits < 5
  insufficient_commits: boolean     // conversations_with_commits < 5
}

export interface WeeklyVerification {
  week: string
  review_before_accept_rate: number
  test_before_commit_rate: number
  edit_clusters: number
  commits: number
  conversations_with_edits: number
  conversations_with_commits: number
  conversation_count: number
}

function roundRate(x: number): number {
  return Math.round(x * 10000) / 10000
}

function isoWeekKey(timestamp: string): string | null {
  const d = new Date(timestamp)
  if (isNaN(d.getTime())) return null
  const dayOfYear = Math.floor((d.getTime() - new Date(d.getFullYear(), 0, 1).getTime()) / 86400000) + 1
  const dayOfWeek = d.getDay() || 7
  const weekNum = Math.ceil((dayOfYear - dayOfWeek + 10) / 7)
  let year = d.getFullYear()
  if (weekNum < 1) {
    year--
    return `${year}-W52`
  }
  if (weekNum > 52) {
    const dec31 = new Date(year, 11, 31)
    const dec31Day = dec31.getDay() || 7
    if (dec31Day < 4) return `${year + 1}-W01`
  }
  return `${year}-W${String(weekNum).padStart(2, '0')}`
}

function flattenTools(messages: TimestampedMessage[]): { tools: string[]; commands: string[] } {
  const tools: string[] = []
  const commands: string[] = []
  for (const msg of messages) {
    if (!msg.tool_names) continue
    for (let i = 0; i < msg.tool_names.length; i++) {
      tools.push(msg.tool_names[i])
      commands.push(msg.bash_commands?.[i] || '')
    }
  }
  return { tools, commands }
}

/**
 * Analyze a conversation's tool sequence for two verification patterns:
 * - Review after edit: each maximal Edit/Write cluster is "reviewed" if a
 *   Read/Grep follows within WINDOW tool positions.
 * - Test before commit: each `git commit` Bash is "tested" if a test-runner
 *   Bash precedes it within WINDOW tool positions.
 *
 * Edit clusters are counted as units (not per individual edit) so a long
 * burst of edits followed by a single Read scores as one reviewed cluster
 * rather than inflating both numerator and denominator.
 */
export function computeConversationVerification(
  messages: TimestampedMessage[],
): ConversationVerification {
  const { tools, commands } = flattenTools(messages)

  let editClusters = 0
  let reviewedClusters = 0
  let i = 0
  while (i < tools.length) {
    if (!EDIT_TOOLS.has(tools[i])) {
      i++
      continue
    }
    let end = i
    while (end + 1 < tools.length && EDIT_TOOLS.has(tools[end + 1])) end++
    editClusters++
    const limit = Math.min(end + 1 + WINDOW, tools.length)
    for (let j = end + 1; j < limit; j++) {
      if (REVIEW_TOOLS.has(tools[j])) {
        reviewedClusters++
        break
      }
    }
    i = end + 1
  }

  let commits = 0
  let testedCommits = 0
  for (let k = 0; k < tools.length; k++) {
    if (tools[k] !== 'Bash') continue
    if (!COMMIT_PATTERN.test(commands[k])) continue
    commits++
    const lo = Math.max(0, k - WINDOW)
    for (let m = k - 1; m >= lo; m--) {
      if (tools[m] !== 'Bash') continue
      if (COMMIT_PATTERN.test(commands[m])) break
      if (TEST_PATTERN.test(commands[m])) {
        testedCommits++
        break
      }
    }
  }

  return {
    edits_count: editClusters,
    reviewed_edits: reviewedClusters,
    commits_count: commits,
    tested_commits: testedCommits,
    review_before_accept_rate: editClusters > 0 ? roundRate(reviewedClusters / editClusters) : 0,
    test_before_commit_rate: commits > 0 ? roundRate(testedCommits / commits) : 0,
  }
}

export function computeVerificationMetrics(
  conversations: ParsedConversation[],
): AggregateVerification {
  let totalEdits = 0
  let totalReviewed = 0
  let totalCommits = 0
  let totalTested = 0
  let convsWithEdits = 0
  let convsWithCommits = 0

  for (const conv of conversations) {
    totalEdits += conv.edits_count || 0
    totalReviewed += conv.reviewed_edits || 0
    totalCommits += conv.commits_count || 0
    totalTested += conv.tested_commits || 0
    if ((conv.edits_count || 0) > 0) convsWithEdits++
    if ((conv.commits_count || 0) > 0) convsWithCommits++
  }

  return {
    total_edit_clusters: totalEdits,
    total_reviewed_edits: totalReviewed,
    review_before_accept_rate: totalEdits > 0 ? roundRate(totalReviewed / totalEdits) : 0,
    total_commits: totalCommits,
    total_tested_commits: totalTested,
    test_before_commit_rate: totalCommits > 0 ? roundRate(totalTested / totalCommits) : 0,
    conversations_with_edits: convsWithEdits,
    conversations_with_commits: convsWithCommits,
    conversation_count: conversations.length,
    insufficient_edits: convsWithEdits < MIN_CONVS_FOR_RATE,
    insufficient_commits: convsWithCommits < MIN_CONVS_FOR_RATE,
  }
}

export function computeWeeklyVerification(
  conversations: ParsedConversation[],
): WeeklyVerification[] {
  if (conversations.length === 0) return []

  const weekGroups = new Map<string, ParsedConversation[]>()
  for (const conv of conversations) {
    if (!conv.started_at) continue
    const week = isoWeekKey(conv.started_at)
    if (!week) continue
    const group = weekGroups.get(week)
    if (group) group.push(conv)
    else weekGroups.set(week, [conv])
  }

  const result: WeeklyVerification[] = []
  for (const [week, weekConvs] of weekGroups) {
    const edits = weekConvs.reduce((s, c) => s + (c.edits_count || 0), 0)
    const reviewed = weekConvs.reduce((s, c) => s + (c.reviewed_edits || 0), 0)
    const commits = weekConvs.reduce((s, c) => s + (c.commits_count || 0), 0)
    const tested = weekConvs.reduce((s, c) => s + (c.tested_commits || 0), 0)
    const convsWithEdits = weekConvs.filter(c => (c.edits_count || 0) > 0).length
    const convsWithCommits = weekConvs.filter(c => (c.commits_count || 0) > 0).length

    result.push({
      week,
      review_before_accept_rate: edits > 0 ? roundRate(reviewed / edits) : 0,
      test_before_commit_rate: commits > 0 ? roundRate(tested / commits) : 0,
      edit_clusters: edits,
      commits,
      conversations_with_edits: convsWithEdits,
      conversations_with_commits: convsWithCommits,
      conversation_count: weekConvs.length,
    })
  }
  result.sort((a, b) => a.week.localeCompare(b.week))
  return result
}

File: webapp/verification_behaviors.py
"""CodeFluent — Verification behavior detection in tool sequences.

Mirrors vscode-extension/src/verificationBehaviors.ts. Detects two patterns:
- Review after edit: a maximal Edit/Write cluster is "reviewed" when a
  Read or Grep follows within WINDOW tool positions.
- Test before commit: a `git commit` Bash is "tested" when a test-runner
  Bash precedes it within WINDOW tool positions, with the lookback
  truncated at the previous commit so a test for an earlier commit
  doesn't credit a later one.
"""

import re
from datetime import datetime

WINDOW = 10
MIN_CONVS_FOR_RATE = 5
EDIT_TOOLS = {"Edit", "Write", "NotebookEdit"}
REVIEW_TOOLS = {"Read", "Grep"}
TEST_PATTERN = re.compile(
    r"\b(npm (run )?test|pytest|jest|cargo test|go test|rspec|mocha|vitest|bun test|tox)\b",
    re.IGNORECASE,
)
COMMIT_PATTERN = re.compile(r"\bgit commit\b")


def _round_rate(x: float) -> float:
    return round(x * 10000) / 10000


def _get_iso_week_key(timestamp: str) -> str | None:
    try:
        dt = datetime.fromisoformat(timestamp.replace("Z", "+00:00"))
        iso = dt.isocalendar()
        return f"{iso.year}-W{iso.week:02d}"
    except (ValueError, AttributeError):
        return None


def _flatten_tools(messages: list[dict]) -> tuple[list[str], list[str]]:
    tools: list[str] = []
    commands: list[str] = []
    for msg in messages:
        names = msg.get("tool_names") or []
        cmds = msg.get("bash_commands") or []
        for i, name in enumerate(names):
            tools.append(name)
            commands.append(cmds[i] if i < len(cmds) else "")
    return tools, commands


def compute_conversation_verification(messages: list[dict]) -> dict:
    """Compute per-conversation review-after-edit and test-before-commit rates."""
    tools, commands = _flatten_tools(messages)

    edit_clusters = 0
    reviewed_clusters = 0
    i = 0
    while i < len(tools):
        if tools[i] not in EDIT_TOOLS:
            i += 1
            continue
        end = i
        while end + 1 < len(tools) and tools[end + 1] in EDIT_TOOLS:
            end += 1
        edit_clusters += 1
        limit = min(end + 1 + WINDOW, len(tools))
        for j in range(end + 1, limit):
            if tools[j] in REVIEW_TOOLS:
                reviewed_clusters += 1
                break
        i = end + 1

    commits = 0
    tested_commits = 0
    for k in range(len(tools)):
        if tools[k] != "Bash" or not COMMIT_PATTERN.search(commands[k]):
            continue
        commits += 1
        lo = max(0, k - WINDOW)
        for m in range(k - 1, lo - 1, -1):
            if tools[m] != "Bash":
                continue
            if COMMIT_PATTERN.search(commands[m]):
                break
            if TEST_PATTERN.search(commands[m]):
                tested_commits += 1
                break

    return {
        "edits_count": edit_clusters,
        "reviewed_edits": reviewed_clusters,
        "commits_count": commits,
        "tested_commits": tested_commits,
        "review_before_accept_rate": _round_rate(reviewed_clusters / edit_clusters) if edit_clusters > 0 else 0,
        "test_before_commit_rate": _round_rate(tested_commits / commits) if commits > 0 else 0,
    }


def compute_verification_metrics(conversations: list[dict]) -> dict:
    total_edits = 0
    total_reviewed = 0
    total_commits = 0
    total_tested = 0
    convs_with_edits = 0
    convs_with_commits = 0

    for conv in conversations:
        total_edits += conv.get("edits_count", 0)
        total_reviewed += conv.get("reviewed_edits", 0)
        total_commits += conv.get("commits_count", 0)
        total_tested += conv.get("tested_commits", 0)
        if conv.get("edits_count", 0) > 0:
            convs_with_edits += 1
        if conv.get("commits_count", 0) > 0:
            convs_with_commits += 1

    return {
        "total_edit_clusters": total_edits,
        "total_reviewed_edits": total_reviewed,
        "review_before_accept_rate": _round_rate(total_reviewed / total_edits) if total_edits > 0 else 0,
        "total_commits": total_commits,
        "total_tested_commits": total_tested,
        "test_before_commit_rate": _round_rate(total_tested / total_commits) if total_commits > 0 else 0,
        "conversations_with_edits": convs_with_edits,
        "conversations_with_commits": convs_with_commits,
        "conversation_count": len(conversations),
        "insufficient_edits": convs_with_edits < MIN_CONVS_FOR_RATE,
        "insufficient_commits": convs_with_commits < MIN_CONVS_FOR_RATE,
    }


def compute_weekly_verification(conversations: list[dict]) -> list[dict]:
    if not conversations:
        return []

    week_groups: dict[str, list[dict]] = {}
    for conv in conversations:
        started_at = conv.get("started_at")
        if not started_at:
            continue
        week = _get_iso_week_key(started_at)
        if not week:
            continue
        week_groups.setdefault(week, []).append(conv)

    result = []
    for week, week_convs in week_groups.items():
        edits = sum(c.get("edits_count", 0) for c in week_convs)
        reviewed = sum(c.get("reviewed_edits", 0) for c in week_convs)
        commits = sum(c.get("commits_count", 0) for c in week_convs)
        tested = sum(c.get("tested_commits", 0) for c in week_convs)
        convs_with_edits = sum(1 for c in week_convs if c.get("edits_count", 0) > 0)
        convs_with_commits = sum(1 for c in week_convs if c.get("commits_count", 0) > 0)
        result.append({
            "week": week,
            "review_before_accept_rate": _round_rate(reviewed / edits) if edits > 0 else 0,
            "test_before_commit_rate": _round_rate(tested / commits) if commits > 0 else 0,
            "edit_clusters": edits,
            "commits": commits,
            "conversations_with_edits": convs_with_edits,
            "conversations_with_commits": convs_with_commits,
            "conversation_count": len(week_convs),
        })

    result.sort(key=lambda x: x["week"])
    return result
